rfcs/20190610-standartizing-composite_ops.md

```diff
@@ -0,0 +1,491 @@
+Suppose we have implemented function `my_function(x) = very(complex(implementation(x)))`. 
+Now, if some downstream library would like to support optimized implemetnation of `my_function`,
+all it needs to do is 
+to replace the tensorflow implementation with its own implementation. However, if at 
+any point we need to compute gradient of `my_function(x)`, then to avoid recomputation, 
+the default implementation of gradient would need all intermediate values produced 
+by tensorflow implementation. In this case this would be values of 
+`implementation(x)`, `complex(implementation(x))`, etc. 
+
+This is problematic for two reasons: 
+
+  1) Downstream have dependence on tensorflow's implementation of the composite ops, and for instance can't just choose arbitrary implementation
+  2)  If tensorflow implementation changes so needs downstream's.
+
+### Option 1: Stabilize Gradient Signature
+Option 1 basically revolves about allowing the composite_op is to provide explicit list of 
+side outputs that could possibly be required for the efficient gradient computation. 
+The `tf.function` machinery would then validate that the signature actually matches 
+the implementation and just respect the order provided in the signature.
+
+This option means that we would not be able to significantly change the implementation in
+the future. For instance, if it is discovered that `my_function(x)` can be computed as 
+`simpler(transformation(x))` we won't be able to change  the implementation it without 
+changing the signature. 
+
+Note, for a non-trivial subset of fast functions this gradient signature could be empty. 
+In fact, *any* gradient  could be computed without any side outputs, by recomputing 
+the function internally. Thus, side outputs only become important when the function involve 
+non-trivial compute.
+
+Thus, this option might be acceptable in simplest cases, where there are no standard outputs, or if side-outputs are extremely standardized. 
+
+
+### Option 2: Keep gradient out of function signature
+Consider the `inference` and `forward` functions. The former has signature `x->y`, the latter is `x->y,  s`.  
+Where either y or s can be a list of multiple tensors.  Comparing the signature, tooling can select
+the split point to separate inference_part and  side output for backward part
+
+>Assumption:  side_outputs of forward_XXX are only ever used as inputs to inference_backward_XXX,
+>if they are used for any other purposes, then the downstream can’t replace the underneath XXX 
+>until those uses are eliminated. (Possibly via graph transformations).  This makes sense because 
+>the graph depends on implementation detail of tf.function implementation, thus tf.function shouldn’t
+>be optimized away.```
+
+Suppose the tooling have an efficient implementation of  our functins that needs its own side outputs let them be t1 ... t_k. Then it can replace all three functions with the re-implementations with the following signatures.
```

Comment:
```suggestion
Suppose the tooling have an efficient implementation of our functions that needs its own side outputs let them be t1 ... t_k. Then it can replace all three functions with the re-implementations with the following signatures.
```